src/main/java/nextstep/subway/line/domain/Sections.java

```diff
@@ -42,106 +44,114 @@ public void addSection(Section section) {
-            sections.add(section);
-        } else if (isFirstSection(searchedSection)) {
-            sections.add(FIRST_INDEX, section);
-        } else {
-            throw new SectionNotSearchedException();
-        }
+        addFirstSection(section);
+        addMiddleSections(section);
+        addLastSection(section);
     }
 
     public void removeSection(Station station) {
-        validateRemoveSections(station);
-        sections.stream()
-                .filter(it -> it.getDownStation().equals(station))
-                .findFirst()
-                .ifPresent(it -> sections.remove(it));
+        if (sections.size() <= SINGLE_SIZE) {
+            throw new EmptyLineException();
+        }
+        List<Section> searchedSections = searchSections(station);
+        sections.removeAll(searchedSections);
+        if (searchedSections.size() > SINGLE_SIZE) {
+            sections.add(mergeSections(searchedSections, station));
+        }
     }
 
     public List<Station> getStations() {
         if (sections.isEmpty()) {
-            return Arrays.asList();
+            return Collections.emptyList();
         }
 
         List<Station> stations = new ArrayList<>();
-        Station downStation = findUpStation();
-        stations.add(downStation);
+        Map<Station, Station> downMapper = sections.stream()
+                .collect(Collectors.toMap(Section::getUpStation, Section::getDownStation));
 
+        Station downStation = getFirstUpStation();
         while (downStation != null) {
-            Station finalDownStation = downStation;
-            Optional<Section> nextLineStation = sections.stream()
-                    .filter(it -> it.getUpStation() == finalDownStation)
-                    .findFirst();
-            if (!nextLineStation.isPresent()) {
-                break;
-            }
-            downStation = nextLineStation.get().getDownStation();
             stations.add(downStation);
+            downStation = downMapper.get(downStation);
         }
 
         return stations;
     }
 
-    private Section searchSection(Section section) {
-        return sections.stream()
-                .filter(section::search)
-                .findAny()
-                .orElseThrow(SectionNotSearchedException::new);
+    private List<Section> searchSections(Station... stations) {
+        List<Section> searchedSections = sections.stream()
+                .filter(section -> Arrays.stream(stations).anyMatch(section::hasStation))
+                .collect(Collectors.toList());
+        if (searchedSections.isEmpty()) {
+            throw new SectionNotSearchedException();
+        }
+        return searchedSections;
+    }
+
+    private Section mergeSections(List<Section> sections, Station station) {
+        Section upSection = sections.stream().filter(
+                section -> station.equals(section.getDownStation())
+        ).findAny().orElseThrow(SectionMergeFailedException::new);
```

Comment:
```suggestion
        Section upSection = sections.stream()
                .filter(section -> station.equals(section.getDownStation()))
                .findAny()
                .orElseThrow(SectionMergeFailedException::new);
```
이 부분 개행 처리를 빠뜨리신것 같아요!